### PR TITLE
Implement soft auto provisioning

### PR DESCRIPTION
**Pull request by julien-nc**
Problem: When auto provisioning is enabled, if a user with the same ID exists in another backend (Db, LDAP...), we create a second user in the user_oidc backend which leads to issues and confusion.

Here is a partial solution to this problem:
New config `soft_auto_provision` flag (enabled by default).
* When enabled
    * If the user already exists in another backend, we don't create a new one in our backend. We update the information (mapped attributes) of the existing user.
    * If the user does not exist in another backend, we create it in ours
* When disabled
    * We refuse login of users that already exist in other backends

This does not migrate existing users to the user_oidc backend but allows them to log in without creating a duplicate user.

Nothing was changed when auto provisioning is disabled.

Extra change: In the user backend, the ldap sync was not triggered when looking for an existing user.

### CODE REVIEW

File: lib/Service/ProvisioningService.php [hunk 1]
```diff
@@ -93,17 +95,21 @@ public function provisionUser(string $tokenUserId, int $providerId, object $idTo
...
+					$this->userMapper->update($backendUser);
+				}
+				// 2 reasons why we should update the display name: It does not match the one
+				// - of our backend
+				// - returned by the user manager (outdated one before the fix in https://github.com/nextcloud/user_oidc/pull/530)
+				if ($newDisplayName !== $oldDisplayName || $newDisplayName !== $user->getDisplayName()) {
+					$this->eventDispatcher->dispatchTyped(new UserChangedEvent($user, 'displayName', $newDisplayName, $oldDisplayName));
+				}
+			} else {
+				$user->setDisplayName($newDisplayName);
```
Comment (nc-fkl) [inline lib/Service/ProvisioningService.php:112]: Not sure here if the $newDisplayName in L112 is always set as its only set once when the if condition in L101 was triggered

**Review by nc-fkl — CHANGES_REQUESTED** [review]
Please check my comment regarding $newDisplayName in ProvisioningService.php

**Review by nc-fkl — APPROVED** [review]
Nice! :)

**Comment by julien-nc** [comment]
@juliushaertl Could you check if this PR makes sense? :grin:

**Review by juliushaertl — APPROVED** [review]
The approach looks good to me, haven't tested though 👍

**Comment by brtptrs** [comment]
Would it be possible to catch this Problem during while doing this provisioning?
https://github.com/nextcloud/user_oidc/issues/656
Maybe with a config flag to enable a lowercase - uppercase verification

**Pull request merged**